src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -968,12 +1004,29 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
                     bb = target;
                 }
                 _ => {
-                    self.not_const();
+                    self.not_const(ops::Loop);
+                    validator.check_op(ops::Loop);
                     break;
                 }
             }
         }
 
+        // The new validation pass should agree with the old when running on simple const bodies
+        // (e.g. no `if` or `loop`).
+        if !use_new_validator {
+            let mut new_errors = validator.take_errors();
+
+            // FIXME: each checker sometimes emits the same error with the same span twice in a row.
+            self.errors.dedup();
+            new_errors.dedup();
+
+            // FIXME: Downgrade this to a warning.
+            if self.errors != new_errors {
+                error!("old validator: {:?}", self.errors);
+                error!("new validator: {:?}", new_errors);
+                panic!("disagreement between validators.");
```

Comment:
Idea by @eddyb
```suggestion
                if tcx.sess.unstable_features.is_nightly_build() {
                	bug!("disagreement between validators.");
                }
```

Reply:
Sounds good. I want to make the panic message more descriptive if we're going to deploy this in the wild and not just in a `crater` run. Also, could we add a temporary `-Z` flag (e.g. `suppress-const-validation-back-compat-ice`) to turn the panic off? Hopefully after they report it :smile:

Reply:
This now causes an ICE only on nightly and only when `-Zsuppress-const-validation-back-compat-ice` is not enabled. The output from the call to `span_bug` looks like this:

```text
error: internal compiler error: src/librustc_mir/transform/qualify_consts.rs:1021: Disagreement between legacy and dataflow-based const validators.
    After filing an issue, use `-Zsuppress-const-validation-back-compat-ice` to compile your code.
   --> src/libcore/cmp.rs:313:12
    |
313 |     Less = -1,
    |            ^^
```

Reply:
Great!